python/adjoint/objective.py

```diff
@@ -132,4 +133,149 @@ def get_evaluation(self):
+                    for xi in range(len(self.dg.x)):
+                        scale = -dJ_4d[:,xi,yi,zi] * dV * 1j * 2 * np.pi * self.frequencies / np.array([self.time_src.fourier_transform(f) for f in self.frequencies])
+                        src = FilteredSource(self.time_src.frequency,self.frequencies,scale,dt,self.time_src)
+                        self.sources += [mp.Source(src, component=self.component, amplitude= 1,
+                                center=mp.Vector3(self.dg.x[xi], self.dg.y[yi], self.dg.z[zi]))]
+
+        return self.sources
+
+    def __call__(self):
+        self.time_src = self.sim.sources[0].src
+
+        self.dg = Grid(*self.sim.get_array_metadata(dft_cell=self.monitor))
+        self.eval = np.array([self.sim.get_dft_array(self.monitor, self.component, i) for i in range(self.num_freq)]) #Shape = (num_freq, [pts])
+        return self.eval
+
+    def get_evaluation(self):
+
+        try:
+            return self.eval
+        except AttributeError:
+            raise RuntimeError("You must first run a forward simulation.")
+
+
+
+class Far_Coefficients(ObjectiveQuantitiy):
+    def __init__(self,sim,Near2FarRegions, far_pt):
+        self.sim = sim
+        self.Near2FarRegions=Near2FarRegions
+        self.eval = None
+        self.far_pt = far_pt
+        return
+
+    def register_monitors(self,frequencies):
+        self.frequencies = np.asarray(frequencies)
+        self.num_freq = len(self.frequencies)
+        self.monitor = self.sim.add_near2far(self.frequencies, *self.Near2FarRegions, yee_grid=True)
+        return self.monitor
+
+    def place_adjoint_source(self,dJ):
+
+        dt = self.sim.fields.dt
+
+        if self.sim.cell_size.y == 0:
+            dV = 1/self.sim.resolution
+        elif self.sim.cell_size.z == 0:
+            dV = 1/self.sim.resolution * 1/self.sim.resolution
+        else:
+            dV = 1/self.sim.resolution * 1/self.sim.resolution * 1/self.sim.resolution
+
+        self.sources = ['src_vol']
+
+
+        freq_scale = 1j * 2 * np.pi * self.frequencies / np.array([self.time_src.fourier_transform(f) for f in self.frequencies])
+
+        dJ = list(dJ.flatten())
+        dJ_c = mp.ComplexVector(len(dJ))
+        for i in range(len(dJ)):
+            dJ_c[i] = dJ[i]
+
+        #TODO far_pts in 3d or cylindrical, perhaps py_v3_to_vec from simulation.py
+        self.all_nearsrcdata = self.monitor.swigobj.near_sourcedata(mp.vec(self.far_pt.x, self.far_pt.y), dJ_c)
+
+
+        for near_data in self.all_nearsrcdata:
+            cur_comp = near_data.near_fd_comp
+            amp_arr = near_data.amp_arr.reshape(self.num_freq, -1)
+
+
+            if cur_comp in [mp.Ex, mp.Ey, mp.Ez]:
+                scale = -freq_scale * amp_arr
+            else:
+                scale = freq_scale * amp_arr
+
+            if self.num_freq == 1:
+                ##TODO
+                self.sources += [mp.Source(self.time_src, component=cur_comp, amplitude=scale[0], center=near_pt)]
```

Comment:
Seems like this should be an array of amplitudes for each spatial point, not a single point source?